Serialized request execution sample (#18)
scalar-labs/scalardl-node-client-sdk: example/serialized_request/README.md

# A Guide on How to Execute a Serialized Request in Client SDKs
This guide briefly explains how to execute a serialized request in client SDKs by looking at how Node SDK can execute a request created by Web SDK.

## Prerequisite
The diagram below shows the whole architecture of this guide.

![](img/samplecode.png)

> Web client is the client side (for example browser or mobile devices) of a product/service that is

Reviewer: Sorry to say this but I don't think it's well written since probably mixing up what they are, what they do in this guide and some requirements that this sample code can work.

> integrated with ScalarDLT, normally we would serialize the requests into one binary array (Uint8Array)

Reviewer: Overall, please use `Scalar DLT` instead of `ScalarDLT`. (a space is needed)

and send to a designated backend service (Web service). Web service is the backend of the product/service where all the
business logic are executed here. In this guide, the web service would just forward the serialized
binary requests from the client side to the scalarDLT server. Therefor the web service and ScalarDLT
server has to be up and running.
Refer here on how to [run a ScalarDLT server](https://scalardl.readthedocs.io/en/latest/) .

`Web-client` and `Web-service` Scalar DL SDKs are initialized as follows:
```
// web-client
const webClientProperties = {
    'scalar.dl.client.cert_holder_id': `foo@`,
    'scalar.dl.client.cert_version': 1,
    'scalar.dl.client.private_key_pem': '...',
    'scalar.dl.client.cert_pem': '...',
  };
const webClient = new Scalar.ClientService(webClientProperties);

// web-service
const properties = {
  'scalar.dl.client.server.host': '127.0.0.1',
  'scalar.dl.client.server.port': 50051,
  'scalar.dl.client.server.privileged_port': 50052,
};
const webService = new ClientServiceWithBinary(properties);
```

> ## Run the sample

Reviewer: Maybe I asked the same thing before, and I think it should describe how to create a request in client side and how to proxy the request in the backend to Scalar DLT, instead of explaining how to run the sample.
Please note that what users of this document want to know is how to create a request and proxy, not how to run the sample (of-course they will run after understanding what it is).

> # Use the node sdk to execute serialized request generated by another application

Reviewer: ```suggestion
# A guide on how to use Node SDK to execute a serialized request created by another application
```

This guide explains how the node sdk can be used as a proxy to execute serialized request.

> To present this feature, we will rely on a demonstration application.

Reviewer: ```suggestion
For ease of explanation, we will use a demo application.
```

> ## Architecture overview

Reviewer: ```suggestion
## Architecture overview of demo application
```

> In this demonstration application, the web client will use the [Scalar DL web-sdk](https://github.com/scalar-labs/scalardl-web-client-sdk) to generate a serialized request then it will

Reviewer: ```suggestion
In the demo application, a web client uses [Scalar DL Web Client SDK](https://github.com/scalar-labs/scalardl-web-client-sdk) to create a serialized request then it
```

> send it to the web service, using the Scalar DL node-sdk, which will execute the request against the Scalar DL server.

Reviewer: ```suggestion
sends it to a web service using Scalar DL Node Client SDK, which executes the request against the Scalar DL network.
```

![overview](img/overview.png)

> ### Serialized request creation using the web sdk

Reviewer: ```suggestion
### Create a serialized request with Web Client SDK
```

> By calling the `clientService.createSerializedXXX` method name, it creates the request as a `Uint8Array`.

Reviewer: ```suggestion
You can create a serialized request as an array of `Uint8Array` with `createSerializedXXX` methods of `ClientService` object as follows.
```

```javascript
const clientProperties = {
    'scalar.dl.client.cert_holder_id': `foo`,
    'scalar.dl.client.cert_version': 1,
    'scalar.dl.client.private_key_pem': '...',
    'scalar.dl.client.cert_pem': '...',
  };
const clientService = new Scalar.ClientService(clientProperties);
const serializedCertificateRegistrationRequest = await clientService.createSerializedCertificateRegistrationRequest();
```

For the full code, please refer to the [source](web-client/index.html) file.

> ### Serialized request execution using the node sdk

Reviewer: ```suggestion
### Execute a serialized request with Node Client SDK
```

> To execute a serialized request, it just needs to be passed as is to its corresponding method.

Reviewer: ```suggestion
You can execute a serialized request with `ClientServiceWithBinary` object as follows.
```

```javascript
const  binaryClientProperties= {
  'scalar.dl.client.server.host': '127.0.0.1',
  'scalar.dl.client.server.port': 50051,
  'scalar.dl.client.server.privileged_port': 50052,
};
const binaryClientService = new ClientServiceWithBinary(binaryClientProperties);
await binaryClientService.registerCertificate(serializedCertificateRegistrationRequest);
```

It should be noted that since the binary client service is only executing the request which was created before by web-client,

> the binary client service does not require the user private key if it only executes serialized request as in a way it only passes the already created request along to the Scalar DL server.

Reviewer: ```suggestion
the binary client service does not require a user private key if it only executes the serialized request.
```
